File: server/server/services/yahoo_finance.py
import os

import requests
from dotenv import load_dotenv
from fastapi import HTTPException
from pydantic import BaseModel, validator
from starlette.status import HTTP_404_NOT_FOUND

load_dotenv()

BASE_URL = os.getenv('YAHOO_API_URL')


class YahooStockModel(BaseModel):
    '''
    TODO: Look into the _ vs . conversion and how to do it better.
    '''
    symbol: str
    currency: str
    regularMarketPrice: float
    previousClose: float

    @validator('symbol')
    def turn_dots_into_underscores(cls, value: str):
        return value.replace('.', '_')

    @classmethod
    def serialize_stock(cls, response_json: dict):
        serialized = response_json \
            .get('chart')          \
            .get('result')[0]      \
            .get('meta')

        return cls.parse_obj(serialized)


def get_stock_info(ticker: str, _range='5d', interval='5m') -> YahooStockModel:
    '''
    TODO: Look into the _ vs . conversion and how to do it better.
    '''
    url = f"{BASE_URL}{ticker.replace('_', '.')}"
    query = {
        'range': _range,
        'interval': interval
    }

    response = requests.get(url, params=query)

    if not response.ok:
        raise HTTPException(
            status_code=HTTP_404_NOT_FOUND,
            detail='Yahoo Finance is currently unavailable.'
        )

    return YahooStockModel.serialize_stock(response.json())

File: pwa/src/services/yahooFinance.ts
import axios, { AxiosResponse } from 'axios'

const YAHOO_API_URL = 'https://query1.finance.yahoo.com/v7/finance/spark'

interface YahooResponseData {
    spark: {
        result: YahooStockResult[]
    }
}

interface YahooStockResult {
    response: [
        { meta: YahooStock }
    ]
}

export interface YahooStock {
    symbol: string
    currency: string
    regularMarketPrice: number
    chartPreviousClose: number
}

const parseResponse = (response: AxiosResponse<YahooResponseData>): YahooStock[] => {
    const { data } = response
    const results = data.spark.result

    // This mapping "enters" the json object simplifying the final result
    const stocksInfo = results.map((result: YahooStockResult) => (result.response[0].meta))

    return stocksInfo
}

export const getStockInfo = async (tickerList: string[]): Promise<YahooStock[]> => {
    const params = {
        symbols: tickerList.join(','),
        range: '1d',
        interval: '1d',
        // corsDomain: 'finance.yahoo.com',
    }

    const response = await axios.get(YAHOO_API_URL, { params })

    return parseResponse(response)
}
